### PR TITLE
LibJS: Loosened typing and miscellaneous changes

### PR DESCRIPTION
**Pull request by 0xtechnobabble**
This patch implements a few miscellaneous things, fixes a few bugs and most importantly, loosens the typing system and implements abstract equality.

### CODE REVIEW

File: Libraries/LibJS/Cell.cpp [hunk 1]
```diff
@@ -40,7 +40,7 @@ void Cell::Visitor::visit(Value value)
         visit(value.as_cell());
 }
 
-Heap& Cell::heap()
+Heap& Cell::heap() const
```
Comment (0xtechnobabble) [inline Libraries/LibJS/Cell.cpp]: I have a feeling like this isn't ok, is this ok?

Comment (0xtechnobabble) [inline Libraries/LibJS/Cell.cpp]: I have a feeling like this isn't ok, is this ok?

File: Libraries/LibJS/AST.h [hunk 2]
```diff
@@ -380,6 +382,34 @@ class StringLiteral final : public Literal {
...
+public:
+    explicit NullLiteral()
+    {
+    }
+
+    virtual Value execute(Interpreter&) const override;
+    virtual void dump(int indent) const override;
+
+private:
+    virtual const char* class_name() const override { return "StringLiteral"; }
```
Comment (oriko1010) [inline Libraries/LibJS/AST.h]:
```suggestion
    virtual const char* class_name() const override { return "NullLiteral"; }
```

File: Libraries/LibJS/AST.h [hunk 3]
```diff
@@ -380,6 +382,34 @@ class StringLiteral final : public Literal {
...
+public:
+    explicit UndefinedLiteral()
+    {
+    }
+
+    virtual Value execute(Interpreter&) const override;
+    virtual void dump(int indent) const override;
+
+private:
+    virtual const char* class_name() const override { return "StringLiteral"; }
```
Comment (oriko1010) [inline Libraries/LibJS/AST.h]:
```suggestion
    virtual const char* class_name() const override { return "UndefinedLiteral"; }
```

File: Libraries/LibJS/StringPrototype.cpp [hunk 4]
```diff
@@ -48,7 +48,7 @@ StringPrototype::StringPrototype()
         ASSERT(this_object);
         i32 index = 0;
         if (!arguments.is_empty())
-            index = arguments[0].to_i32();
+            index = static_cast<i32>(arguments[0].to_number().as_double());
```
Comment (awesomekling) [inline Libraries/LibJS/StringPrototype.cpp]: I don't want to have to do this every time I want to convert a `JS::Value` to an `i32`.
Let's have a helper for it.

File: Libraries/LibJS/Value.cpp [hunk 5]
```diff
@@ -89,113 +87,108 @@ Value Value::to_object(Heap& heap) const
...
+        return Value(m_value.as_double);
+    case Type::Null:
+        return Value(0);
+    case Type::String: {
+        bool ok;
+        //FIXME: Parse in a better way
+        auto parsed_int = as_string()->string().to_int(ok);
+        if (ok)
+            return Value(parsed_int);
+        else
```
Comment (awesomekling) [inline Libraries/LibJS/Value.cpp]: No need for `else` after `return`.

File: Libraries/LibJS/Cell.cpp (hunk 1 above)
Comment (awesomekling) [reply Libraries/LibJS/Cell.cpp]: I think it's ok.

**Comment by awesomekling** [comment]
Great! :)

**Pull request merged**